### ISSUE TITLE
cli: dump doesn't preserve nanoseconds in timestamps

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#8804]
Insert two rows with differing timestamps using `parse_timestamp_ns`. Using `dump` on them drops the nanosecond precision. Need to make sure the recover preserves these too.

```
> create table t (ts timestamp);
CREATE TABLE
root@:26257> insert into t  values (parse_timestamp_ns('2016-08-24 19:28:56.643155000'));
INSERT 1
root@:26257> insert into t  values (parse_timestamp_ns('2016-08-24 19:28:56.643155001'));
INSERT 1
root@:26257> select * from t;
+----------------------------------------+
|                   ts                   |
+----------------------------------------+
| 2016-08-24 19:28:56.643155 +0000 +0000 |
| 2016-08-24 19:28:56.643155 +0000 +0000 |
+----------------------------------------+
(2 rows)
```

```
$ ./cockroach dump d t
CREATE TABLE t (
        ts TIMESTAMP NULL,
        FAMILY "primary" (ts, rowid)
);

INSERT INTO t VALUES
        ('2016-08-24 19:28:56.643155+00:00'),
        ('2016-08-24 19:28:56.643155+00:00');
```

**Comment by mjibson** [comment]
TLDR: nano's aren't in the PG wire spec, just remove them completely

After discussion with @dt and @paperstreet, the three of us now agree that the best resolution to this is to remove nanoseconds from Cockroach timestamps completely. Specifically, we would remove the `parse_timestamp_ns` and `format_timestamp_ns`. The rest of the infrastructure already purges nanos everywhere.

The argument here is that supporting, say, some `SET` option that enables the output of nanos can only work for the text protocol on PG drivers that happen to support nanos (which isn't in the spec). If a user really really wants nanos, they can use a `INT` and figure it out themselves. Adding a `TIMESTAMPNS` type also doesn't make a lot of sense because no driver would know how to speak that type since we would be making it up.

**Comment by mjibson** [comment]
@bdarnell @petermattis could one of you comment about this proposal to remove nanos? We need some more agreement outside of us before moving forward.

**Comment by knz** [comment]
The alternative is:
1) have dump output to nanosecond precision
2) have the text parser choose to either parse microseconds or nanoseconds depending on how many digits are specified
3) have an additional SET for strict pg compatibility

I am personally reluctant to drop features "just because pg doesn't support them". Extensions are fine if the migration path is well-defined. Having compatibility with pg tools constrain what we can/want to support is like a tail wagging a dog.

**Comment by knz** [comment]
(NB: I get that you "talked" about this, but then it'd be nice to have a recap of the discussion here.)

**Comment by mjibson** [comment]
1) The problem isn't dump outputting nanoseconds, it's the pgwire package not sending the nanos. If we start sending nanos, we break spec compat. There isn't even a way to send nanos using the binary protocol, because it's an int of microseconds from an epoch.
2) Having the parser recognize the nano digits is what it used to do, and that's where the original problems came from.
3) Yes, we could do this if we think it's correct. If choose this I like the idea of allowing setting of the timestamp format style explicitly, and the user can choose what they want by reading the Go time docs. However this further complicates our protocol spec compliance because we'd be sending a timestamp as a randomly formatted string, hoping that the driver does the correct thing. Risky and language/driver dependent. Not great.

Discussion recap (not in time order, just trying to remember ideas):

- We have spent a lot of time on this already, and it seems clearer now that dealing with nanos is just a bad idea because postgres doesn't have them in its protocol and we claim to support the postgres protocol.
- Making a new type isn't useful because it'd have to use new OIDs that other drivers don't understand. Although maybe we could just force it to always marshal as a string?
- Durations are currently dealing with this problem too, see #8755.
- We like the idea of keeping nanoseconds because lots of time code has them now, but we can't think of a good way to do it.
- We leave everything as is but document that parse_timestamp_ns produces data that isn't present in `cockroach dump`. And if you happen to have a timestamp with nanos as a PK, you could be in trouble when you restore a backup.
- We could augment `cockroach dump` to learn about timestamps and check if they have nanos (by using `format_timestamp_ns`. This would add a lot of special, ugly logic to the dump subcommand, but would work.

**Comment by knz** [comment]
To react to the individual points:

- Making a new type isn't useful because it'd have to use new OIDs that other drivers don't understand. Although maybe we could just force it to always marshal as a string?
   
yes marshal as string. Pgwire has support for custom and user-defined and plugin-defined types we can piggy back either of these.

- Durations are currently dealing with this problem too, see #8755.

Yes but obviously the two issues are linked and will be addressed together in the same way. 

- We leave everything as is but document that parse_timestamp_ns produces data that isn't present in `cockroach dump`. And if you happen to have a timestamp with nanos as a PK, you could be in trouble when you restore a backup.

What kind of trouble? Can you be specific with an example?

- We could augment `cockroach dump` to learn about timestamps and check if they have nanos (by using `format_timestamp_ns`. This would add a lot of special, ugly logic to the dump subcommand, but would work.

Again we can use the text formatting for this. 6 digits for micros, 9 for nanos.

**Comment by mjibson** [comment]
Another effect this bug has is that when paging through the data, since the nanoseconds are truncated, a row with a timestamp PK will get that row twice if it's at a page boundary.

**Comment by mjibson** [comment]
The trouble is the same as in the original issue description, but assuming the `ts` column is the PK. During insert the two timestamps won't be unique and so there will be an error. The other problem with timestamp as PK is the above comment, where paging is incorrect because of the truncated value, which causes a duplicate row to be fetched.

**Comment by dt** [comment]
relatedly, if you _ever_ wrote a timestamp with non-zero nanos, if you later switch drivers or add add other micro services services to your stack that use the standard wire protocol, you might find yourself fighting subtle and surprising bugs where some of your rows can't be selected (because even if you parrot back the timestamp you read, it doesn't match what's on disk). this seems like the kind of non-obvious and very frustrating db bugs/gotchas I specifically want to avoid inflicting on users.

**Comment by tschottdorf** [comment]
I can't comment on whether supporting nanosecond timestamps is adequate for the decade (I suppose it is), but intermingling it with any of the postgres timestamps seems fraught, judging from my limited experience with this bug.

**Comment by mjibson** [comment]
@knz We were in agreement that if it could be done, adding a new type would be ok if we had requests from users  for nanosecond-precision timestamps. The custom OIDs thing is interesting and something I didn't know, and would likely make this possible.

**Comment by knz** [comment]
NB if we end up removing support for nanoseconds then do not forget to remove the corresponding operators from EXTRACT.
